Add example of data streaming with apache kafka + apache flink (#353)
crate/cratedb-examples: application/apache-kafka-flink-streaming/README.md

# Streaming data with Apache Kafka, Apache Flink and CrateDB.

## About

This example showcases what a data-streaming architecture leveraging Kafka and Flink could look
like.

We use.

- Kafka (confluent)
- Apache Flink
- CrateDB 
- Python >=3.7<=3.11
- 

> [img of the thing]

Reviewer: Either add an image, or remove this line? When I am adding screenshots for those purposes, I am usually not adding them to the content tree in the repository itself. Instead, I am pasting them into a an issue comment on GitHub I will never submit, and then grab the image URL out of the text field after the image upload concluded.

## Overview

An HTTP call is scheduled to run every 60 seconds on `weather_producer`, the API returns a JSON
with the specified city's weather, the json is then sent through `Kafka`.

`flink_consumer` is a flink application consuming the same kafka topic;

> upon receiving data, it sends the resulting datastream to the sink, which is `CrateDB`

Reviewer: ```suggestion
upon receiving data, it sends the resulting datastream to the sink, which is CrateDB.
```

Both `flink_consumer` and `weather_producer` are written using their respective Python Wrappers.

[kafka-python](https://kafka-python.readthedocs.io/en/master/)

[apache-flink](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/python/overview/)

Everything is customizable via environment variables, the API schedule, the topic, credentials...
etc.

> See `.env` more details.

Reviewer: ```suggestion
See `.env` for more details.
```

## How to use

> There is ready-to-use docker-compose, fill in the .env with the API
> Key (Get it here https://www.weatherapi.com/).

Reviewer: ```suggestion
The Docker Compose configuration will get you started quickly.
You will just need to fill in the API key of [Weather API](https://www.weatherapi.com/)
into your local `.env` file.
```

Reviewer 2: :+1:

### Run the docker compose (and build the images)

> `docker compose up -d --build`

Reviewer: ````suggestion
```
docker compose up -d --build
```
````

### Stop the docker compose

> `docker compose down`

Reviewer: ````suggestion
```
docker compose down
```
````

### Poetry

> `poetry install`

Reviewer: ````suggestion
```
poetry install
```
````

### Pip

> `pip install -r requirements.txt`

Reviewer: ````suggestion
```
pip install -r requirements.txt
```
````

## Notes

### CrateDB initial settings.

CrateDB stores the shard indexes on the file system by mapping a file into memory (mmap)
You might need to set `max_map_count` to something higher than the usual default, like `262144`.

You can do it by running `sysctl -w vm.max_map_count=262144`,
for more information see: [this](https://cratedb.com/docs/guide/admin/bootstrap-checks.html#linux)

### Mock API call.

If you don't want to register in the weather api we use, you can use the
provided `mock_fetch_weather_data`, call this instead in the scheduler call.

> This is what it'd look like.

Reviewer 2: ```suggestion
This is how it'd look like:
```

```python
scheduler.enter(
    RUN_EVERY_SECONDS,
    1,
    schedule_every,
    (RUN_EVERY_SECONDS, mock_fetch_weather_data, scheduler)
)
```

> *after changing this, re-build the docker compose.*

Reviewer: Can you also add the command please?

```suggestion
*After changing this, re-build the docker compose.*
```

### Initial kafka topic.

In this example the `Kafka` topic is only initialized the first data is sent to it, because of this
the flink job could fail if it exceeds the default timeout (60) seconds, this might only happen
if the API takes too long to respond *the very first time this project*.

To solve this, you should [configure](https://kafka.apache.org/quickstart#quickstart_createtopic)
the

> topics on boot time. This is recommended for production scenarios.

Reviewer 2: ```suggestion
topics at boot time. This is recommended for production scenarios.
```
